server/monitor.go

```diff
@@ -142,173 +162,209 @@ func (s *Server) Connz(opts *ConnzOptions) (*Connz, error) {
-			ci.Port = addr.Port
-			ci.IP = addr.IP.String()
-		}
-
+		ci := &conns[i]
+		ci.fill(client, client.nc, c.Now)
 		// Fill in subscription data if requested.
-		if subs {
-			sublist := make([]*subscription, 0, len(client.subs))
+		if subs && len(client.subs) > 0 {
+			ci.Subs = make([]string, 0, len(client.subs))
 			for _, sub := range client.subs {
-				sublist = append(sublist, sub)
+				ci.Subs = append(ci.Subs, string(sub.subject))
 			}
-			ci.Subs = castToSliceString(sublist)
 		}
-
 		// Fill in user if auth requested.
 		if auth {
 			ci.AuthorizedUser = client.opts.Username
 		}
-
 		client.mu.Unlock()
+		pconns[i] = ci
+		i++
+	}
+	// Closed Clients
+	for _, cc := range closedClients {
+		// Fill in subscription data if requested.
+		if subs && len(cc.subs) > 0 {
+			cc.Subs = cc.subs
+		}
+		// Fill in user if auth requested.
+		if auth {
+			cc.AuthorizedUser = cc.user
+		}
+		pconns[i] = &cc.ConnInfo
 		i++
 	}
+
+	switch sortOpt {
+	case ByCid:
+		sort.Sort(byCid{pconns})
+	case BySubs:
+		sort.Sort(sort.Reverse(bySubs{pconns}))
+	case ByPending:
+		sort.Sort(sort.Reverse(byPending{pconns}))
+	case ByOutMsgs:
+		sort.Sort(sort.Reverse(byOutMsgs{pconns}))
+	case ByInMsgs:
+		sort.Sort(sort.Reverse(byInMsgs{pconns}))
+	case ByOutBytes:
+		sort.Sort(sort.Reverse(byOutBytes{pconns}))
+	case ByInBytes:
+		sort.Sort(sort.Reverse(byInBytes{pconns}))
+	case ByLast:
+		sort.Sort(sort.Reverse(byLast{pconns}))
+	case ByIdle:
+		sort.Sort(sort.Reverse(byIdle{pconns}))
+	case ByUptime:
+		sort.Sort(sort.Reverse(byCid{pconns}))
+	}
+
+	minoff := c.Offset
+	maxoff := c.Offset + c.Limit
+
+	maxIndex := totalClients
+
+	// Make sure these are sane.
+	if minoff > maxIndex {
+		minoff = maxIndex
+	}
+	if maxoff > maxIndex {
+		maxoff = maxIndex
+	}
+
+	// Now pare down to the requested size.
+	// TODO(dlc) - for very large number of connections we
+	// could save the whole list in a hash, send hash on first
+	// request and allow users to use has for subsequent pages.
+	// Low TTL, say < 1sec.
+	c.Conns = pconns[minoff:maxoff]
+	c.NumConns = len(c.Conns)
+
 	return c, nil
 }
 
+// Fills in the ConnInfo from the client.
+// client should be locked.
+func (ci *ConnInfo) fill(client *client, nc net.Conn, now time.Time) {
+	ci.Cid = client.cid
+	ci.Start = client.start
+	ci.LastActivity = client.last
+	ci.Uptime = myUptime(now.Sub(client.start))
+	ci.Idle = myUptime(now.Sub(client.last))
+	ci.RTT = client.getRTT()
+	ci.OutMsgs = client.outMsgs
+	ci.OutBytes = client.outBytes
+	ci.NumSubs = uint32(len(client.subs))
+	ci.Pending = int(client.out.pb)
+	ci.Name = client.opts.Name
+	ci.Lang = client.opts.Lang
+	ci.Version = client.opts.Version
+	// inMsgs and inBytes are updated outside of the client's lock, so
+	// we need to use atomic here.
+	ci.InMsgs = atomic.LoadInt64(&client.inMsgs)
+	ci.InBytes = atomic.LoadInt64(&client.inBytes)
+
+	// If the connection is gone, too bad, we won't set TLSVersion and TLSCipher.
```

Comment:
I believe you should use `nc` down below, not `client.nc`...

Reply:
Agree will fix.